[Semantics] Parse the AST into a semantically valid structure

c-harding opened the pull request.

c-harding committed:
[Semantics] Convert abstract syntax tree to a meaningful representation

c-harding committed:
[Semantics] Add support for parsing structs

c-harding committed:
[Semantics] Refactor and document semantic parsing

Also rewrite the metadata matcher to store a list of matchers, rather
than nested closures of functions

c-harding committed:
[Semantics] Remove commented config from scalafmt

c-harding commented on src/main/scala/temple/DSL/Semantics/Test.scala:
```diff
@@ -0,0 +1,17 @@
+
+    println(semanticParse)
+  }
+}
```
These need to be converted to real unit tests

c-harding replied:
and Exception can be turned into scalatest's `fail`

c-harding commented on src/main/scala/temple/DSL/Semantics/Metadata.scala:
```diff
@@ -0,0 +1,20 @@
+  // TODO: is string right here? Could these three be enums?
+  case class Language(name: String)             extends GlobalMetadata
+  case class Provider(name: String)             extends ProjectMetadata
+  case class Database(name: String)             extends ProjectAndServiceMetadata
```
This is probably worth it, I'll update as soon as I can

c-harding commented on src/main/scala/temple/DSL/Semantics/SemanticParsingException.scala:
```diff
@@ -0,0 +1,5 @@
+package temple.DSL.Semantics
+
+/** An exception in converting the AST into a [[temple.DSL.Semantics.Templefile]] */
+// TODO: include contextual information
```
Adding this context is definitely useful, but will take a while

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+          )
+        case "bool"   => AttributeType.BoolType
+        case typeName => fail(s"Unknown type $typeName")
+        //TODO: handle annotations
```
I think this is the only part of the AST yet to be processed

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+
+  /** A parser of Metadata items that can occur in project and service blocks */
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
```
I can't add this to a constructor because of the additional type flags, but thoughts on this (versus `inheritFrom(parseGlobalMetadata)`, or another way of expressing it?

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+        tag match {
+          case "struct" => structs.safeInsert(key -> parseStructBlock(entries)(BlockContext(key, tag, context)))
+          case tag      => fail(s"Unknown block type $tag for $key in $context")
+        }
```
Should this be flattened? I guess so, despite the additional repetition of `DSLRootItem(key, _, entries)`

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+        fail(s"Found ${otherEntry.typeName} in ${context.tag} block (${context.block}): `$otherEntry`")
+    }
+    StructBlock(attributes.toMap)
+  }
```
Any thoughts/suggestions on how these can be merged?

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+    var projectNameBlock: Option[(String, ProjectBlock)] = None
+
+    val targets  = mutable.LinkedHashMap[String, TargetBlock]()
+    val services = mutable.LinkedHashMap[String, ServiceBlock]()
```
`LinkedHashMap` used to maintain insertion order, I should add a comment to explain that

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  /** Any element of a service/struct */
+  sealed abstract class Entry(val typeName: String)
```
`typeName` should be added to the ScalaDoc

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,67 @@
+
+    case class Attribute(key: String, dataType: AttributeType, annotations: List[Annotation])
+        extends Entry("attribute") {
+      override def toString: String = s"$key: $dataType${annotations.map(" " + _).mkString};"
```
More string building, seems very like the SQL builder we want to make

xsanda commented on src/main/scala/temple/DSL/package.scala:
```diff
@@ -1,76 +0,0 @@
-      s"$key: $tag {\n$contents\n}"
-    }
-  }
-}
```
Wow, this file changed so much git couldn't keep up

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,67 @@
+  }
+
+  /** An item at the root of the Templefile, e.g. services and targets */
+  sealed case class DSLRootItem(key: String, tag: String, entries: List[Entry]) extends Entry(s"$tag block ($key)") {
```
This `typeName` should be refactored into the new tracing stuff, as at https://github.com/TempleEight/temple/pull/22/files#diff-bea48dd33be35e22f9384f36ccc3cbc1R13

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,67 @@
+    def size: Int        = posargs.size + kwargs.size
+    def isEmpty: Boolean = posargs.isEmpty && kwargs.isEmpty
+
+    override def toString: String = {
```
Should this stuff be the default `toString`?

c-harding commented:
Code coverage low score! Will add testing soon too

c-harding committed:
[Semantics] Remove old mkString override

c-harding committed:
[Semantics] Add support for field annotations

c-harding commented on src/main/scala/temple/DSL/Semantics/package.scala:
```diff
@@ -0,0 +1,383 @@
+import scala.collection.immutable.ListMap
+import scala.collection.mutable
+
+package object Semantics {
```
Should this be a static package object?

c-harding committed:
[Semantics] Rearrange semantics structure to avoid package object

c-harding committed:
[Semantics] Replace uses of List with Seq

List is a linked list, and converting into this data structure seems a
bit pointless in many cases when they can just be left in the Seq
(immutable ordered collection) that they are already in/can be converted
to most efficiently.

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Fix annotation rerendering to Templefile

c-harding committed:
[Semantics] Add tests for SeqUtils

c-harding committed:
[Semantics] Fix zipped deprecation warning

c-harding committed:
[Semantics] Remove superfluous default case in match block

c-harding committed:
[Semantics] Remove local testing Main function

c-harding committed:
Merge branch 'develop' into semantics

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Fix some minor warnings

c-harding committed:
[Semantics] Extract MapUtils to separate file with tests

c-harding committed:
[Semantics] Start to test Semantic Analyser

c-harding commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,384 @@
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("database", TokenArgType, Database)
+  }
```
I can't add this to a constructor because of the additional type parameters, but thoughts on this (versus `inheritFrom(parseGlobalMetadata)`, or another way of expressing it/building up these parsers?

c-harding replied:
@jaylees14 @Lewiky thoughts?

c-harding committed:
Merge branch 'develop' into semantics

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,67 @@
+
+  /** A sequence of values, both positional and named */
+  case class Args(posargs: Seq[Arg] = Nil, kwargs: Seq[(String, Arg)] = Nil) {
+    def size: Int        = posargs.size + kwargs.size
```
This is never called, should I delete it? @jaylees14

c-harding committed:
[Semantics] Test semantic type parsing

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Add tests to semantic analyser

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Document use of LinkedHashMap

c-harding committed:
[Semantics] Add Entry’s parameter to ScalaDoc

c-harding committed:
[Semantics] Use enums for language, provider and database

c-harding commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,395 @@
+        fail(s"Found ${otherEntry.typeName} in ${context.tag} block (${context.block}): `$otherEntry`")
+    }
+    StructBlock(attributes.toMap)
+  }
```
Any thoughts/suggestions on how these can be merged?

c-harding commented on src/main/scala/temple/DSL/Syntax.scala:
```diff
@@ -0,0 +1,69 @@
+import temple.utils.StringUtils.indent
+
+/** Handles the internal representation of the DSL of the Templefile */
+object Syntax {
```
This file comes from `src/main/scala/temple/DSL/package.scala`

c-harding commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,395 @@
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("database", TokenArgType)(Database.parse(_))
+  }
```
I can't add this to a constructor because of the additional type parameters, but thoughts on this (versus `inheritFrom(parseGlobalMetadata)`, or another way of expressing it/building up these parsers?

Lewiky replied:
Agree with Jay on just explicitly starting the complete list of keywords at each bottom level object and taking the repetition hit

c-harding committed:
Merge branch 'syntax-refactor' into semantics

c-harding committed:
[Semantics] Break up semantic analyser to multiple files

c-harding committed:
[Semantics] Refactor parse__Block to avoid repetition

Lewiky commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,263 @@
+  }
+
+  // TODO: is this too messy?
+  def parseAttributeType(dataType: syntax.AttributeType)(implicit keyNameContext: KeyName): AttributeType = {
```
If this is public, probably needs docs

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Fix merge diff commit

Lewiky reviewed (commented):
Would be nice to have some docs

c-harding committed:
[Semantics] Fix parseMetadataBlock type error

c-harding committed:
[Semantics] Add documentation to parseAttributeType

jaylees14 commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,264 @@
+  /** A parser of Metadata items that can occur in target blocks */
+  private val parseTargetMetadata = new MetadataParser[TargetMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("auth", "services", ListArgType(TokenArgType))(TargetAuth)
```
Personally, I don't like the `inherit from` here. Feels a bit like spaghetti code.
I think it'd make more sense to avoid the indirection and explicitly register each keyword with the target/project/service metadata parser every time

c-harding committed:
Merge remote-tracking branch 'origin/develop' into semantics

c-harding committed:
[Semantics] Refactor metadata parser to remove inheritance

c-harding committed:
[Semantics] Fix missing language metadata at project level

c-harding committed:
Merge branch 'develop' into semantics

c-harding committed:
Merge pull request #35 from TempleEight/semantics-refactor-metadata-parser

[Semantics] Refactor metadata parser to remove inheritance

c-harding committed:
[Semantics] Revert TODO of possible test

c-harding commented on src/main/scala/temple/DSL/semantics/Analyser.scala:
```diff
@@ -0,0 +1,262 @@
+    * @return A parsed attribute type
+    */
+  // TODO: is this too messy?
+  def parseAttributeType(dataType: syntax.AttributeType)(implicit keyNameContext: KeyName): AttributeType = {
```
@jaylees14 thoughts on this function? Any way to rewrite it?

jaylees14 commented on src/main/scala/temple/DSL/semantics/AttributeType.scala:
```diff
@@ -0,0 +1,20 @@
+trait AttributeType
+
+object AttributeType {
+  case class ForeignKey()  extends AttributeType
```
Plz be object

c-harding committed:
[Semantics] Resolve PR comments

Make an empty case class into an object, and remove a todo that no
longer needs to be done

Lewiky reviewed (approved):
LGTM
